### ISSUE TITLE
Errors when syncing to S3Path directory that doesn't exist

### ISSUE THREAD

**Issue by nickrobinson251** [JuliaCloud/AWSS3.jl#76]
For comparison, here's the `aws-cli` behaviour:
```sh
$ aws s3 sync path/to/dir s3://mybucket-13puwo3rvcn01/does/not/exist --dryrun
(dryrun) upload: path/to/dir/file1.jl to s3://mybucket-13puwo3rvcn01/does/not/exist/file1.jl
(dryrun) upload: path/to/dir/file2.jl to s3://mybucket-13puwo3rvcn01/does/not/exist/file2.jl
```
```sh
$ aws s3 sync help

DESCRIPTION
       Syncs  directories  and S3 prefixes. Recursively copies new and updated
       files from the source directory to the destination. Only creates  fold-
       ers in the destination if they contain one or more files.
```
```sh
$ aws --version
aws-cli/1.17.3 Python/3.7.3 Darwin/17.7.0 botocore/1.14.3
```

Compare the behaviour with `S3Path`:

1) without trailing slash:
```julia
julia> local_path = abs(p"path/to/dir");

julia> s3path = p"s3://mybucket-13puwo3rvcn01/does/not/exist";

julia> sync(local_path, s3path)
ERROR: ArgumentError: S3Path folders must end with '/': s3://mybucket-13puwo3rvcn01/does/not/exist
```
^ i'm not sure if we want this behaviour or not. I think `S3Path` is consistent on this, but it is inconsistent with `aws-cli`.

Anyway, this bring us on to:

2) with trailing slash
```julia
julia> s3path_ = p"s3://mybucket-13puwo3rvcn01/does/not/exist/";

julia> sync(local_path, s3path_)
ERROR: The parent of s3://gpfexperiments-outputbucket-13puwo3rvcn01/does/not/exist/Experiments.jl/src/ does not exist. Pass `recursive=true` to create it.
```
^ This is think is just a bug?

**Comment by nickrobinson251** [comment]
I notice we add a method to [`sync(f::Function, src::AbstractPath, dst::S3Path)`](https://github.com/JuliaCloud/AWSS3.jl/blob/7e5e227ee9bf813582e259c5bcf668ad8000a9e5/src/s3path.jl#L217)
so although we initially hit [`sync(src::AbstractPath, dst::AbstractPath)`](https://github.com/rofinn/FilePathsBase.jl/blob/v0.7.0/src/path.jl#L532), that should then redirect us to the 3-arg `sync` method in this package.

In any case, i believe when `isdir(s3path) == true` and `exists(s3path) == false)`, then [`sync` just fallsback to `cp`](https://github.com/JuliaCloud/AWSS3.jl/blob/v0.6.9/src/s3path.jl#L234-L273).

And, in `cp`, if  `dest` (`s3path`) doesn't exist [it first calls `mkdir(dest)`](https://github.com/rofinn/FilePathsBase.jl/blob/v0.7.0/src/path.jl#L494). And [`mkdir(dest::S3path)`](https://github.com/JuliaCloud/AWSS3.jl/blob/v0.6.9/src/s3path.jl#L234-L273) throws (both) the errors we see above.

So i think [`FilePathsBase.cp`](https://github.com/rofinn/FilePathsBase.jl/blob/v0.7.0/src/path.jl#L494) should instead call `mkdir(dest; recursive=true)`.

If this is the fix, I do not know if we want to make it here, with a `cp(src, dest::S3Path` method, or upstream in FilePathsBase for all `dest::AbstractPath`.

**Comment by nickrobinson251** [comment]
any thoughts @mattBrzezinski @rofinn ?

**Comment by rofinn** [comment]
The current behaviour is consistent with how base `cp` works. We should not change that.

```julia
julia> mktempdir() do d
           mkpath(joinpath(d, "foo", "bar"))
           touch(joinpath(d, "foo", "bar", "baz.txt"))
           cp(joinpath(d, "foo"), joinpath(d, "fred", "plugh"))
           readdir(joinpath(d, "fred", "plugh"))
       end
ERROR: SystemError (with /var/folders/vz/zx_0gsp9291dhv049t_nx37r0000gn/T/jl_KOlhud/fred/plugh): mkdir: No such file or directory
Stacktrace:
 [1] #systemerror#44(::String, ::typeof(systemerror), ::Symbol, ::Bool) at ./error.jl:134
 [2] #systemerror at ./none:0 [inlined]
 [3] #mkdir#7(::UInt16, ::typeof(mkdir), ::String) at ./file.jl:175
 [4] mkdir at ./file.jl:170 [inlined]
 [5] #cptree#11(::Bool, ::Bool, ::typeof(Base.Filesystem.cptree), ::String, ::String) at ./file.jl:311
 [6] #cptree at ./none:0 [inlined]
 [7] #cp#12(::Bool, ::Bool, ::typeof(cp), ::String, ::String) at ./file.jl:342
 [8] cp at ./file.jl:338 [inlined]
 [9] (::var"#8#9")(::String) at ./REPL[17]:4
 [10] #mktempdir#18(::String, ::typeof(mktempdir), ::var"#8#9", ::String) at ./file.jl:634
 [11] mktempdir(::Function, ::String) at ./file.jl:632 (repeats 2 times)
 [12] top-level scope at REPL[17]:1
```
```

**Comment by nickrobinson251** [comment]
> The current behaviour is consistent with how base cp works. We should not change that.

Great point. i agree :)

> Isn't the issue solved by simply creating the non-existent directory ahead of time?

This is indeed a work around.

i'd still consider `sync(src, dest::S3Path)` having different behaviour to `aws s3 sync` to be surprising / a "gotcha". We can choose to mitigate the gotcha via docs, but i'd prefer to just match aws-cli. I suspect we can do that without chaning the `cp` behaviour in general.

**Comment by rofinn** [comment]
The issue is that `aws s3 cp` is also differs from base `cp`, so it seems weird to have `sync` work that way, but not `cp`. I could see adding `s3sync` and `s3cp` if you want the cli behaviour.

**Comment by nickrobinson251** [comment]
> I could see adding s3sync and s3cp if you want the cli behaviour.

does `s3_copy` have the behaviour of `aws s3 cp`? Sorry for my ignorance here, it's been a long time since I used AWSS3 functionality other than S3Path.

**Comment by rofinn** [comment]
> does s3_copy have the behaviour of aws s3 cp?

No, `s3_copy` refers to the [rest call](http://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectCOPY.html) which will make a copy of a single object already stored in S3. Similar to how `s3_put` only uploads a single object with a given prefix.

**Comment by nickrobinson251** [comment]
Thanks :)

> it seems weird to have `sync` work that way, but not `cp`.

I was on the fence about this, but after our offline conversation I a leaning towards agreeing. 

I think this is issue is best resolved via improved docs. 

Given that `sync` is not a Julia base command, and `aws s3 sync` is a thing, I suggest we add a note to the `sync(src, dest::S3Path)` docstring pointing out the does-not-create-destination behaviour, noting it matches `cp` but differs from `aws s3 sync`, and perhaps directing folks to `mkdir`.

**Comment by nickrobinson251** [comment]
xref https://github.com/rofinn/FilePathsBase.jl/issues/65 (thanks, Rory!)

**Comment by nickrobinson251** [comment]
> it seems weird to have sync work that way, but not cp.

Okay, I am back to thinking this is not weird. And the best solution is to have `sync` have the `mkdir(dest; recursive=true)` behaviour, because you can just use `cp` otherwise.

**Comment by rofinn** [comment]
> because you can just use cp otherwise.

That's `cp` and `sync` are not the same thing. `sync` does proper file synchronization, only updating files at the destination that have changed (by default it considers size and last modified date). If you don't need the sync functionality then I recommend using `mkdir` with `cp` instead... at it'll be faster.
